Add a schema loader for CakePHP core tests

markstory opened the pull request:
I was planning on using phinx and a single phinx migration for this, however that is not going to work. Phinx depends on cakephp/database which cannot be installed as it would conflict with the root package.

Because of that limitation I've added a way to use the Database package's schema generation features more directly. Retaining this long term will term will require having some schema generation capabilities in CakePHP. Those features could be extracted out into an internal package for 5.x.

Related to #15308

markstory committed:
Add a schema loader for CakePHP core tests

I was planning on using phinx and a single phinx migration for this,
however that is not going to work. Phinx depends on cakephp/database
which cannot be installed as it would conflict with the root package.

Because of that limitation I've added a way to use the Database
package's schema generation features more directly. Retaining this long
term will term will require having some schema generation capabilities
in CakePHP. Those features could be extracted out into an internal package
for 5.x.

markstory commented on tests/schema.php:
```diff
@@ -0,0 +1,12 @@
+ * features of the Database package.
+ */
+return [
+];
```
I'm planning to generate the contents of this file using the existing fixture definitions, but I'll do that in a separate pull request.

raul338 commented:
Would the code from [cakephp-schema](https://github.com/raul338/cakephp-schema) be useful here?

Currently generates a `schema.php` with all db tables and uses to fill the fixutres `$schema`

markstory committed:
Fix variable shadowing and move success message so it isn't duplicated.

markstory commented:
> Would the code from cakephp-schema be useful here?
> 
> Currently generates a schema.php with all db tables and uses to fill the fixutres $schema

Perhaps for the generation step. With the [new fixtures plan](https://github.com/cakephp/cakephp/issues/15308) we're planning to move away from schema in fixture files. Instead applications would rely on migrations (or SQL dumps) to create schema for tests to reduce maintenance toil. Longer term this should allow us to remove schema generation features from CakePHP's public APIs which also helps simplify the database code.

markstory committed:
Fix psalm error

raul338 commented:
Yes, with that plan and the [FixtureFactories](https://github.com/vierge-noire/cakephp-fixture-factories) plugin, the cakephp-schema would be deprecated.

Feel free to copy the contents of the [SchemaLoadCommand](https://github.com/raul338/cakephp-schema/blob/master/src/Command/SchemaLoadCommand.php) (I think you already have your implementation) and [SchemaSaveCommand](https://github.com/raul338/cakephp-schema/blob/master/src/Command/SchemaSaveCommand.php)

markstory committed:
Create schema with constraints disabled.

markstory commented on tests/bootstrap.php:
```diff
@@ -139,3 +140,7 @@
 // has been written to.
 session_id('cli');
+
+// Create test database schema.
```
Yes it would.

ADmad replied:
That totally sucks. Can't the schema generation be done only when it's actually needed?

ADmad replied:
Plugins tests uses this bootstrap file https://github.com/cakephp/bake/blob/b3b4d1206cd6fbe7b647f67669627f0414b94e42/templates/bake/Plugin/tests/bootstrap.php.twig#L51.

So with currently proposed change even those plugins that have nothing to do with the database would be unnecessarily generating the schema.

markstory replied:
> That totally sucks. Can't the schema generation be done only when it's actually needed?

Yes. I'm thinking that most plugins will use migrations based schema, and that interface would not reset schema on each run.

We could 'skip' dropping tables that already exists. This would mean that core developers would need to manually wipe their test db when existing fixtures are changed.

Do you want to see how much of an impact clearing and rebuilding the test schema for cakephp's test takes before making a firm decision?

ADmad replied:
> If a plugin is pulling in our tests/bootstrap.php they should probably not do that as it isn't a 'public' interface.

Well they do, the plugin's tests/bootstap.php generated by bake includes the core's tests/bootstrap.php as seen in the template I linked above. It's a good convenience and you don't have to repeat the boiler plate code in each plugin. 

Could we perhaps have a phpunit config based, on which the schema generator usage would be skipped? Then one can either set that config in phpunit.xml or pass at command line.

othercorey replied:
Not to derail the review, but does that means tests/bootstrap.php *is* considered a public api and cannot change in patch releases? I don't think I've seen migration notes on changes to it -- like when we use DB_URL.

markstory replied:
It is in the gray area of public/internal it seems. I'd prefer it wasn't public but there will likely be lots of plugins that do depend on it to define constants and paths.

markstory committed:
Don't rebuild schema in isolated tests.

Use a potentially hacky sniff to detect if we are process isolated and
skip creating the database schema.

markstory committed:
Fix table ordering.

markstory committed:
Revert change to query logging

markstory committed:
Only use fixture schema if an env var is set

This will allow new schema to be used in core tests, but not in plugins.
I'd prefer if plugins didn't use schema metadata, but it is likely going
to happen.

ADmad commented on src/TestSuite/Schema/SchemaGenerator.php:
```diff
@@ -0,0 +1,103 @@
+     */
+    public function reload(?array $tables = null): void
+    {
+        if (!file_exists($this->file)) {
```
Why not do this check in constructor itself since there's no other public interface to modify this property?

markstory replied:
Fair point.

markstory committed:
Move file validation into the constructor.

markstory committed:
Add index and rename foreign key.

markstory committed:
Add missing index.

markstory commented:
Merging so I can continue the next block of work which is converting the existing fixture schema to the new format, and switching the core tests to use the new schema tooling.
